plugins/che-plugin-analytics/che-plugin-analytics-wsagent/src/main/java/com/redhat/che/plugin/analytics/wsagent/AnalyticsManager.java

```diff
@@ -0,0 +1,434 @@
+                    dispatcher.getUserId());
+              } else {
+                synchronized (dispatcher) {
+                  AnalyticsEvent lastEvent = dispatcher.getLastEvent();
+                  if (lastEvent == null) {
+                    return;
+                  }
+
+                  long expectedDuration = lastEvent.getExpectedDurationSeconds() * 1000;
+                  if (lastEvent == WORKSPACE_INACTIVE
+                      || (expectedDuration >= 0
+                          && System.currentTimeMillis()
+                              > expectedDuration + dispatcher.getLastEventTime())) {
+                    if (dispatcher.sendTrackEvent(
+                            WORKSPACE_USED,
+                            Collections.emptyMap(),
+                            dispatcher.getLastIp(),
+                            dispatcher.getLastUserAgent())
+                        != null) {
+                      return;
+                    }
+                  }
+                }
+              }
+
+              networkExecutor.submit(dispatcher::sendPingRequest);
+            });
+  }
+
+  public void onActivity(String userId) {
+    try {
+      dispatchers.get(userId).onActivity();
+    } catch (ExecutionException e) {
+      LOG.warn("", e);
+    }
+  }
+
+  public void onEvent(
+      String userId,
+      AnalyticsEvent event,
+      Map<String, Object> properties,
+      String ip,
+      String userAgent) {
+    if (event == WORKSPACE_STARTED) {
+      workspaceStartingUserId = userId;
+    }
+    try {
+      dispatchers.get(userId).sendTrackEvent(event, properties, ip, userAgent);
+    } catch (ExecutionException e) {
+      LOG.warn("", e);
+    }
+    ;
+  }
+
+  private class EventDispatcher {
+
+    private String userId;
+    private String cookie;
+
+    private AnalyticsEvent lastEvent = null;
+    private Map<String, Object> lastEventProperties = null;
+    private long lastActivityTime;
+    private long lastEventTime;
+    private String lastIp = null;
+    private String lastUserAgent = null;
+
+    private Map<String, Object> commonProperties;
+
+    EventDispatcher(String userId, AnalyticsManager manager) {
+      this.userId = userId;
+      this.commonProperties =
+          ImmutableMap.<String, Object>of(
+              WORKSPACE_ID, workspaceId,
+              WORKSPACE_NAME, workspaceName);
+      this.cookie =
+          Hashing.md5()
+              .hashString(workspaceId + userId + System.currentTimeMillis(), StandardCharsets.UTF_8)
+              .toString();
+      LOG.info(
+          "Analytics Woopra Cookie for user {} and workspace {} : {}", userId, workspaceId, cookie);
+    }
+
+    void onActivity() {
+      lastActivityTime = System.currentTimeMillis();
+    }
+
+    void sendPingRequest() {
+      try {
+        URI uri =
+            new URI(
+                "http://www.woopra.com/track/ping?host="
```

Comment:
Please extract constant here and for the parameters

Reply:
fixed